shared object - ref impl with java/jni

jrhea opened the pull request:
Resolves #20

@vyzo @bigs @raulk I am not totally done, but it might be worth taking a look at what i have done so far to make sure i am coloring in the lines. 

> Note: I am a golang neophyte so i apologize in advance for anything that is no idiomatic.  Also, let me know if the liberties i took with project/code organization need to be changed.

### Build Instructions:
- install packages dependencies `make deps`
- to build the daemon for go `make` or `make go-daemon`
- to build the client for go `make go-client`
- to build the daemon for java `make java-daemon`
- to build the daemon for java `make java-client`

### Modifications 

complete:
- added build option for an .so and generation of jni source for java (osx only)
- ~added a client option to main.go~ made a separate executable for the client called p2pc
- added some additional functionality to the goclient to test the java daemon
- added .gitignore
- add go and java subfolders to p2pclient
- control the daemon/client Go code from Java
- cleanup the domain socket file if client fails
- .so build command for linux
- created .so for: daemon only, client only and daemon/client combined

<h3 id="functionality">Functionality</h4>
i can fire up two daemons on two different domain sockets:

**The first one is run from java**

```bash
$ java p2pd
Control socket: /tmp/p2pd.sock
Peer ID: Qmb3mEWb7JGBxzduizmFnzQ14uYaqsFqmsCU4tssMRY3rT
Peer Addrs:
/ip6/::1/tcp/61459
/p2p-circuit
/ip4/127.0.0.1/tcp/61458
/ip4/192.168.1.129/tcp/61458
/ip4/192.168.2.65/tcp/61458 
```
**the second one from go:**

```bash
$ p2pd --sock=/tmp/p2pd2.sock
Control socket: /tmp/p2pd2.sock
Peer ID: Qmbr1NfbBMAzhWTxsJUYu1oXXHhSVRx9BW13qR8uMXuhio
Peer Addrs:
/ip6/::1/tcp/59344
/p2p-circuit
/ip4/127.0.0.1/tcp/59343
/ip4/192.168.1.129/tcp/59343
/ip4/192.168.2.65/tcp/59343
```
Then I can run two clients to interact with each daemon respectively:

**client 1 - listen for message**
```bash
$ java p2pc --command=ListenForMessage 
```

**client 2 - connect to other daemon then send message**
```bash
p2pc  --command=Connect  --pathc=/tmp/p2c2.sock --pathd=/tmp/p2pd2.sock Qmb3mEWb7JGBxzduizmFnzQ14uYaqsFqmsCU4tssMRY3rT /ip4/127.0.0.1/tcp/61458

p2pc --command=SendMessage --pathc=/tmp/p2c2.sock --pathd=/tmp/p2pd2.sock Qmb3mEWb7JGBxzduizmFnzQ14uYaqsFqmsCU4tssMRY3rT FOOOOOBAAAAAAARR
```
The message `FOOOOOBAAAAAAARR` is displayed by client 1

>JAVA USERS: you might need to set the java lib path `-Djava.library.path=.`

vyzo commented on Makefile:
```diff
@@ -1,10 +1,43 @@
+java-daemon: libp2pd.jnilib
+
+libp2pd.jnilib : libp2pd.o
+	ld -dylib -flat_namespace -undefined suppress -macosx_version_min 10.13.4 -o $(DIR)/$@ $(DIR)/*.o -L$(DIR) -lp2pd
```
can we find a solution that works on linux as well? Alternate, have explicit targets that call the rules with the macosx name.

mhchia replied:
Is there a specific reason that `macosx_version_min 10.13.4` is set? I have upgraded to 10.14 so extra warnings showed up when I run this.

mhchia replied:
Sorry for the stupid question(I only compiled shared libs in linux): is `libp2pd.jnilib` compiled with `fPIC`? I checked there is a GOT section, but not so sure that it means it is compiled as PIC. Thank you!

jrhea replied:
> Sorry for the stupid question(I only compiled shared libs in linux): is libp2pd.jnilib compiled with fPIC? I checked there is a GOT section, but not so sure that it means it is compiled as PIC. Thank you!

It's not a stupid question. I had to be careful about how I ordered the build steps. If you look at the libp2p.o target you will see that I use $CFLAGS var for the gcc options. I defined those at the top of the file and -fPIC is one of them

jrhea replied:
> can we find a solution that works on linux as well? Alternate, have explicit targets that call the rules with the macosx name.

@vyzo I def plan on doing that.  i will have the Makefile inspect uname to determine what OS to build for

vyzo commented on daemon.go:
```diff
@@ -51,6 +55,15 @@ func NewDaemon(ctx context.Context, path string, opts ...libp2p.Option) (*Daemon
 	go d.listen()
 
+	sigc := make(chan os.Signal, 1)
+	signal.Notify(sigc, os.Interrupt, syscall.SIGTERM)
```
is this actually useful? SIGTERM already terminates the program.

jrhea replied:
@vyzo 
in daemon.go, i added this code:
```
	sigc := make(chan os.Signal, 1)
	signal.Notify(sigc, os.Interrupt, syscall.SIGTERM)
	go func(ln net.Listener, c chan os.Signal) {
		sig := <-c
		fmt.Printf("Caught signal %s: shutting down.", sig)
		ln.Close()
		os.Exit(0)
	}(d.listener, sigc)
```
so that the listener is properly closed (and removes the /tmp/p2pd.sock file) in these two cases:
1) when the user hits Ctrl-C  (sends SIGINT)
2) when this code:
```
//export stopD
func stopD() {
	p, _ := os.FindProcess(os.Getpid())
	p.Signal(os.Interrupt)
}
```
is called.  Basically, I was just sending a SIGINT when stopD is called so that the code in daemon.go is called.  

I am sure you have a better way to do it, but I needed some way to "gracefully" exit

vyzo commented on p2pclient/go/p2pclient.go:
```diff
@@ -30,7 +30,7 @@ type Client struct {
 // listening on a unix socket at controlPath, and establishing an inbound socket
 // at listenPath.
-func NewClient(controlPath, listenPath string) (*Client, error) {
+func NewClient(controlPath string, listenPath string) (*Client, error) {
```
please don't change this, it's idiomatic go to write arguments with common type like this.

jrhea replied:
10-4 Thanks for the clarification

vyzo commented on p2pd/main.go:
```diff
@@ -1,28 +1,128 @@
+		startD()
+	}
+}
+func startC(config ClientConfig) {
```
Please make a separate binary for the client; you can make a directory in the top level (call it `p2pc`) and add a main program in there.

jrhea replied:
> Please make a separate binary for the client; you can make a directory in the top level (call it p2pc) and add a main program in there.

Will do...it was bothering me too :)

mhchia commented on Makefile:
```diff
@@ -1,10 +1,43 @@
+p2pd.h : p2pd.class
+	cd $(DIR) && javac -h $(CLASS_PATH) $*.java
+
+p2pd.class: go-daemon
```
Will this be more efficient to avoid running `deps`?
```suggestion
p2pd.class: $(DIR)/p2pd.java
```

jrhea replied:
> Will this be more efficient to avoid running deps?

@mhchia your suggestion will work if all the go source has been downloaded (e.g. github.com/libp2p/go-libp2p-connmgr), but if it is the first time you are pulling it down...I believe it will fail.  That being said, I understand the spirit of the suggestion and I will modify the build to only conditionally perform those steps.

jrhea commented:
@vyzo & @mhchia almost done here.  These are the final tasks (for this PR):
- .so build command for linux
- finish parsing code to pass all command line params from java client to go

i am leaning towards packaging 
- the shared object 
- the java proxy class that interacts with the shared object
into an API that exposes all the functionality needed to implement the beacon chain.  this would save me the trouble of implementing all the p2pclient code in Java.  this might be useful for other beacon chain implementations that don't have libp2p implemented in their project's language.  

Thoughts?

vyzo commented:
that sounds reasonably -- looking good so far.

raulk commented on p2pc/main.go:
```diff
@@ -0,0 +1,121 @@
+
+	}
+
+	os.Remove(*config.pathc)
```
you should be able to use a `defer` here.

raulk commented on p2pc/main.go:
```diff
@@ -0,0 +1,121 @@
+		log.Fatal(err)
+	}
+
+	if *config.command == "Identify" {
```
Prefer to use a switch over the string.

jrhea replied:
i have several other fixes and cleanup to add.  do you mind if i open another PR ore should I fix these here?

raulk replied:
Better if you can fix it here :-)

jrhea replied:
done

jrhea commented:
@vyzo @raulk @mhchia I believe i addressed all the issues for this PR.  Here's a screen cap of it in action:

![libp2p-daemon-demo](https://user-images.githubusercontent.com/5555162/49203987-eb55fa80-f36f-11e8-9e5e-f94ceddc3117.gif)

vyzo commented on Makefile:
```diff
@@ -1,10 +1,72 @@
-bin: deps
-	go install ./...
+SHELL := /bin/bash
```
we should probably use `/bin/sh` as `/bin/bash` is not available in all systems.

vyzo commented on daemon.go:
```diff
@@ -51,6 +55,15 @@ func NewDaemon(ctx context.Context, path string, opts ...libp2p.Option) (*Daemon
+	signal.Notify(sigc, os.Interrupt, syscall.SIGTERM)
+	go func(ln net.Listener, c chan os.Signal) {
+		sig := <-c
+		fmt.Printf("Caught signal %s: shutting down.\n", sig)
```
let's use the `log` for this. 
So `log.Debugf(...)`.

vyzo commented on p2pd/main.go:
```diff
@@ -1,54 +1,93 @@
+	connMgrGrace   *int
+	QUIC           *bool
+	natPortMap     *bool
+	args           []string
```
this will need rebasing, we also have pubsub arguments now.

raulk commented on p2pclient/java/p2pc.java:
```diff
@@ -0,0 +1,21 @@
+public class p2pc {
```
What is the purpose of having the Java binding in this repo? Is it just sample code, or do we want downstream users to use this as a dependency? If the latter is the case, we should add a Java build system (Maven/gradle) and publish to maven.

raulk replied:
The thing about the Go binding is that it's usable for downstream users in this form. They just import the `...p2pclient/go` package and that's it. But in the current form, this is not usable for Java users.

jrhea replied:
> What is the purpose of having the Java binding in this repo? Is it just sample code

Good question.  I don't want to pollute the repo with unnecessary code.  I can refactor and make it example code if you all prefer that.

jrhea replied:
> The thing about the Go binding is that it's usable for downstream users in this form. They just import the ...p2pclient/go package and that's it. But in the current form, this is not usable for Java users.

Agreed.  Another option is to move the java specific binding code to a bindings/java folder.  Then I could move the java main to an examples/java folder.

jrhea replied:
@raulk I added all code related to bindings to a bindings folder.  Also, i moved all java specific binding code to bindings/java (the example java classes are in bindings/java/examples).  I think this will help

raulk commented on p2pc/main.go:
```diff
@@ -0,0 +1,146 @@
+}
+
+//export startClient
+func startClient(args *C.char) {
```
I'm confused. This is not a _client_ -- this starts an embedded daemon, right? To me a client is a stub that consumes an existing service.

jrhea replied:
Sorry for the confusion here.  So there are two main.go files:

- p2pd/main.go - which starts the daemon
- p2pc/main.go which only starts a client  

The client can send commands to a daemon and listen/wait for messages from the daemon.  If you look at the section titled [Functionality](#functionality) in the first comment in this issue, I have a detailed explanation of the difference.  

I wanted to avoid recreating all the code in /p2pclient in java (and other languages).  It was convenient enough to just expose the go client methods needed to develop the essential functionality for the beacon chain.  That being said, I am not done, but I paused development bc I saw that you guys were hard at work on adding functionality to the daemon still.  Maybe it would make more sense to reorganize the code some.  I will do that today (in addition to making the changes suggested by @vyzo ).

jrhea commented:
@vyzo & @raulk I resolved all the issues and reorganized the code I added into a bindings folder.  The dir structure now looks like:

> -project root
> --bindings
>  ---java
>  ----examples

If you type `make all` it will create:
- p2pd: The daemon executable
- p2pc: The client executable
- libp2pd.dynlib: The daemon shared object with only startDaemon and stopDaemon exported
- libp2pc.dynlib: The client shared object with only startClient exported
- libp2p.dynlib: Combine (daemon & client) shared object with startDaemon, stopDaemon and startClient exported.  ( I made this so that you don't have to import both .so's)

Please take a look and let me know what you think.

vyzo commented on test/mock_daemon.go:
```diff
@@ -8,7 +8,7 @@ import (
 	ggio "github.com/gogo/protobuf/io"
 	"github.com/gogo/protobuf/proto"
-	"github.com/libp2p/go-libp2p-daemon/p2pclient"
+	"github.com/libp2p/go-libp2p-daemon/p2pclient/go"
```
is this still the correct package?

jrhea replied:
sorry, fixing it now

vyzo commented on test/utils.go:
```diff
@@ -13,7 +13,7 @@ import (
 	crypto "github.com/libp2p/go-libp2p-crypto"
 	p2pd "github.com/libp2p/go-libp2p-daemon"
-	"github.com/libp2p/go-libp2p-daemon/p2pclient"
+	"github.com/libp2p/go-libp2p-daemon/p2pclient/go"
```
same here.

jrhea commented:
@vyzo thanks for the review.  I pushed the changes.  I have more functionality to implement in the client side, specifically in [p2pclient/p2pc.go](https://github.com/libp2p/go-libp2p-daemon/blob/aa4e06c9ec47a3c23201d013cac53efc6bbb4fd0/p2pclient/p2pc.go)

The code only supports these commands:

>	Identify
>	Connect
>	ListenForMessage (NewStreamHandler)
>	SendMessage (NewStream)

I was holding off on implementing more functionality until you guys OK this mode of operation.  That highlights one negative aspect of this mode...going forward there will potentially be one more touch point when functionality is added to the daemon.  We will need to make sure it is made available to the shared object.  That being said, the other option would be to create a "LibP2P Client Library" where this is housed.  Thoughts?

vyzo commented:
I am fine with this housed in the daemon repo -- @bigs @raul thoughts?

bigs commented:
yep i would prefer it live within the repo. will do a thorough review tomorrow

jrhea commented:
Thanks @bigs .  Sorry in advance for the pain in the ass this review will be :scream:

jrhea commented:
@bigs have you been able to take a look at this PR?

jrhea commented:
@raulk @mhchia @vyzo @bigs @mvid I drew up some diagrams to share with you guys on the two different ways people can choose to integrate with the libP2P daemon.  I am hoping this will clarify what I was thinking when I created this PR.

![Option 1](https://user-images.githubusercontent.com/5555162/50185987-0963aa80-02df-11e9-8741-5a5b523a5f7a.jpg)


![Option 2](https://user-images.githubusercontent.com/5555162/50186004-14b6d600-02df-11e9-9fba-42a5eda9b16e.jpg)

raulk commented:
Hey @jrhea – FINALLY I had time to review this in-depth! Thank you so much for spearheading this effort in the community 🎉 

Being only vaguely familiar with Cgo, I would've expected the native interface to mimic the client's public interface. In my head, the design would look like this:

1. An application in Java, Ruby, C#... calls a `void start_client()` function on the native interface when starting. This creates a singleton Go client internally and holds a reference to it. We never return the client via the native interface; the return type is void.
2. We have functions `dht_find_peer()`, `pubsub_subscribe()`, `connect()`, etc. that mirror the Go client's interface and proxy over the call to it, transforming between C types and Go types where needed.
3. We might need some kind of `read_stream(int id)` for the application to poll streams. This keeps the IPC behind closed doors and delivers data via the native interface, mimicking the socket API.
4. When the app shuts down, it calls `void stop_client()` method that closes the singleton Go client. We'll probably need a `int status()` method to deterministically enquire the native layer about the status of the Go singleton client.

How does that sound, @jrhea? Happy to collaborate to push this forward.

jrhea commented:
> Hey @jrhea – FINALLY I had time to review this in-depth! Thank you so much for spearheading this effort in the community 🎉

All good my man.  I was happy to do it.  I appreciate you guys letting me participate.

> An application in Java, Ruby, C#... calls a void start_client() function on the native interface when starting. This creates a singleton Go client internally and holds a reference to it. We never return the client via the native interface; the return type is void.

Totally agree and that is exactly what I would like to discuss on our call.  I just implemented the startClient piece with 'magic strings' as commands to prove it works.   When I started the PR I could see that the daemon was still early in development and I didn't want to waste too much time nailing down an interface until you guys were ready.  I figured that once you guys were ready,  we could all agree on what this interface would look like.

> We have functions dht_find_peer(), pubsub_subscribe(), connect(), etc. that mirror the Go client's interface and proxy over the call to it, transforming between C types and Go types where needed.

Cool. I think that structs are the only type that might be a problem (unless they are defined in C), but I think we can work around this issue.

> We might need some kind of read_stream(int id) for the application to poll streams. This keeps the IPC behind closed doors and delivers data via the native interface, mimicking the socket API.

Yes, exactly.  That is probably the key thing to work out.  The Java or C# side could either poll for a message received that matches the stream id or it might be cool to just call an externally defined method from Go like ext_stream_rvcd(int id) - sort of like a callback - and that could avoid polling.

> When the app shuts down, it calls void stop_client() method that closes the singleton Go client. We'll probably need a int status() method to deterministically enquire the native layer about the status of the Go singleton client.

Perfect!

> How does that sound, @jrhea? Happy to collaborate to push this forward.

I like your brain.  I think we are on the same page!  I really think there is tremendous value in creating a rich interface to other languages so that must of the heavy lifting is done on the Go side.  That being said, some teams (e.g. @mhchia ) have already picked the IPC endpoint as their integration point.  That is fine though.  That is why I had the Makefile create two versions of the shared library:

- libP2Pd.so (for @mhchia ) that only exposes the `control interface`
    - startDaemon(config)
    - stopDameon()
- libP2P.so that exposes the `command & control interface`:
    - startDaemon(config)
    - stopDameon()
    - startClient(command)  

> Note: startClient(command) will be changed to a richer interface that you (@raulk ) suggested

jrhea commented:
> mad drawing skillz!

Lol thanks!  Who said developers can't draw?  Amiright? 🎨

raulk commented:
@jrhea yesterday I updated the target branch for this PR to the `native-api` long-lived branch, as discussed. We can keep making progress there, until the point where we're confident that the design is stable, and we then we can merge to master.

## Notes from the call

**Attendees:** @jrhea @vyzo @bigs @mkalinin @Nashatyrev @akhila-raju @schroedingerscode @raulk 

We see this moving forward as a set of discrete PRs:

1. One for simple start(config)/stop functions that start an embedded daemon which exposes IPC endpoints. The app interacts with the daemon via IPC. This is what the team wanting to run the daemon on iOS suggested (I think it was somebody from Status.im, but can't remember who).
2. Laying the foundation for a fully-fledged native API, see https://github.com/libp2p/go-libp2p-daemon/pull/28#issuecomment-448639670. Implementing basic functions like connect, identify.
3. One PR adding native API calls per subsystem (DHT, conn manager, pubsub, etc.)

We discussed concurrency models, i.e. how to expose native API methods whose Go client counterparts deal with channels. Callbacks and event loops were two suggested models. @jrhea to open issue to discuss.

Instead of creating one glue/translation layer (.so lib) between native API <=> Go client, we can create one per each concurrency model we want to support.

Anything else you guys want to add?

jrhea commented:
@raulk @vyzo @bigs 

> Anything else you guys want to add?

Great chatting with you guys!  I just submitted the first PR in @raulk's list: 

- https://github.com/libp2p/go-libp2p-daemon/pull/54

>  Callbacks and event loops were two suggested models. @jrhea to open issue to discuss.

I opened the issue and it can be found: 

- https://github.com/libp2p/go-libp2p-daemon/issues/52

Looking forward to this. Thanks!

bigs commented:
sorry this took a thousand years. looking good!

raulk commented:
@jrhea – IIUC we should close this PR in favour of #54, correct?

jrhea commented:
@raulk ya totally.  This PR should just be closed and not merged.  #54 supersedes it and I will make a new PR to create a native function call interface for native command and native control.
